pkg/infrastructure/gcp/clusterapi/iam.go

```diff
@@ -0,0 +1,113 @@
+	resourcemanager "google.golang.org/api/cloudresourcemanager/v1"
+	iam "google.golang.org/api/iam/v1"
+)
+
+// ServiceAccount predefined roles. Roles are described here https://cloud.google.com/iam/docs/understanding-roles#predefined_roles
+var saRoles = []string{
+	"roles/compute.instanceAdmin",
+	"roles/compute.networkAdmin",
+	"roles/compute.securityAdmin",
+	"roles/storage.admin",
+}
+
+func defaultServiceAccountID(infraID, projectID, role string) string {
+	return fmt.Sprintf("%s-%s@%s.iam.gserviceaccount.com", infraID, role[0:1], projectID)
+}
+
+// CreateServiceAccount is used to create a service account for a compute.
+func CreateServiceAccount(ctx context.Context, infraID, projectID, role string) (string, error) {
+	ctx, cancel := context.WithTimeout(ctx, time.Minute*1)
+	defer cancel()
+
+	service, err := iam.NewService(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to create IAM service: %w", err)
+	}
+
+	accountID := defaultServiceAccountID(infraID, projectID, role)
+	displayName := fmt.Sprintf("%s-%s-node", infraID, role)
+
+	request := &iam.CreateServiceAccountRequest{
+		AccountId: accountID,
+		ServiceAccount: &iam.ServiceAccount{
+			Description: "The service account used by the instances.",
+			DisplayName: displayName,
+		},
+	}
+
+	_, err = service.Projects.ServiceAccounts.Create("projects/"+projectID, request).Do()
+	if err != nil {
+		return "", fmt.Errorf("Projects.ServiceAccounts.Create: %w", err)
+	}
+
+	// TODO - do we need to poll until the resource is found?
+
+	logrus.Debugf("Service account created for %s", accountID)
+
+	return accountID, nil
+}
+
+// AddServiceAccountRoles adds predefined roles for service account.
+func AddServiceAccountRoles(ctx context.Context, projectID, serviceAccountID string) error {
+	policy, err := getProjectIAMPolicy(ctx, projectID)
+	if err != nil {
+		return err
+	}
+
+	// TODO - would be useful to test getting serviceAccount account roles before and after
+
+	for _, role := range saRoles {
+		err = addMemberToRole(policy, role, serviceAccountID)
+		if err != nil {
+			return fmt.Errorf("failed to add role %s to %s: %w", role, serviceAccountID, err)
+		}
+	}
+
+	return nil
+}
+
+func getProjectIAMPolicy(ctx context.Context, projectID string) (*resourcemanager.Policy, error) {
+	ctx, cancel := context.WithTimeout(ctx, time.Minute*1)
+	defer cancel()
+	req := &resourcemanager.GetIamPolicyRequest{}
+
+	service, err := resourcemanager.NewService(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create resourcemanager service: %w", err)
+	}
+
+	policy, err := service.Projects.GetIamPolicy(projectID, req).Context(ctx).Do()
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch project IAM policy: %w", err)
+	}
+	return policy, nil
+}
+
+// addMemberToRole adds a member to a role binding.
+func addMemberToRole(policy *resourcemanager.Policy, role, member string) error {
+	for _, binding := range policy.Bindings {
+		if binding.Role != role {
```

Comment:
Suggest reversing the logic. There can only be one matching role (at most).

Reply:
Changed.